README.md
# Express-Versioned-Route

Simple express.js 4.x extension, to add support for versioned routes.

[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![codecov](https://codecov.io/gh/swvl/express-versioned-route/branch/master/graph/badge.svg)](https://codecov.io/gh/swvl/express-versioned-route)
[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=sqale_rating)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=security_rating)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=alert_status)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)

[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=sqale_index)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=vulnerabilities)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=bugs)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=code_smells)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Duplicated Lines (%)](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=duplicated_lines_density)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)
[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=swvl_express-versioned-route&metric=ncloc)](https://sonarcloud.io/dashboard?id=swvl_express-versioned-route)

## Highlights

- Versioning the same route
- Loading time validation
- Custom middlewares for each version
- Built-in soft deprecation policy
- Case insensitive name matching
- Minimal overhead ( less than 0.15ms/request )
- Dynamic routing based on client headers
- Client can choose specific version using `accept-version` header

## installation
- Update your .npmrc file: `echo "@swvl:registry=https://npm.pkg.github.com" >> .npmrc`
- Then install the package: `npm install @swvl/express-versioned-route`

## Usage

```js
const searchVersionDef = versionsDef()({
  versions: {
    simpleSearch: ['2020-Q1', searchMW1, searchMW2, searchHandlerV1],
    dynamicSearch: [dynamicSearchMW1, dynamicSearchMW2, searchHandlerV2, 'default'],
    superSearch: [searchHandlerV3],
    deepSearch: [searchHandlerV4],
  },
  Android: [
    [400, 'simpleSearch'],
    [450, 'dynamicSearch'],
    [500, 'superSearch'],
  ],
  iOS: [
    [400, 'simpleSearch'],
    [460, 'dynamicSearch'],
  ],
});

app.get('/search', mw1, mw2, searchVersionDef);
```

#### Expected user headers

* device-os & build-number
* accept-version

[![Try it live on RunKit](https://badge.runkitcdn.com/@swvl/express-versioned-route.svg)](https://runkit.com/swvl/express-versioned-route)

## Options

Global options can be passed as 1st parameter for versionsDef function, or configured as part of versionsDef specs

#### Example
```js
const searchVersionDef = versionsDef(globalOptions)({
  versions: {
    ...
  }
  ...,
  options: {
    ...
  }
});
```

#### Supported options

<table>
<tr>
  <td> <b> prop name </b> </td>
  <td> <b> description </b> </td>
  <td> <b> default </b> </td>
</tr>
  
<tr>
  <td> onDeprecated </td>
  <td> callback  </td>
  <td>
    <pre>
(versionName, req)=> {} 
    </pre>
  </td>
</tr>

<tr>
  <td> allowClientRequestFallbackToDefaultVersion </td>
  <td> invalid client headers will fallback to default version instead of 404 </td>
  <td> true </td>
</tr>

<tr>
  <td> onClientSelection </td>
  <td> configure how client should be selected based on req </td>
  <td> 
    <pre>
onClientSelection: req => {
  return {
    name: req.headers['device-os'],
    buildNumber: req.headers['build-number'],
  };
}
    </pre>
  </td>
</tr>
</table>


## License

[Apache-2.0](<https://tldrlegal.com/license/apache-license-2.0-(apache-2.0)>)


CHANGELOG.md
## [3.0.3](https://github.com/swvl/express-versioned-route/compare/v3.0.2...v3.0.3) (2021-12-13)


### Bug Fixes

* minor code smells ([08a2406](https://github.com/swvl/express-versioned-route/commit/08a2406d5c0f072a00da595b8ed6cceb9dcdb222))

## [3.0.2](https://github.com/swvl/express-versioned-route/compare/v3.0.1...v3.0.2) (2021-12-13)

## [3.0.1](https://github.com/swvl/express-versioned-route/compare/v3.0.0...v3.0.1) (2021-12-13)

# [3.0.0](https://github.com/swvl/express-versioned-route/compare/v2.2.4...v3.0.0) (2021-12-13)


### chore

* **package manager:** migration from npm to github ([c939a1e](https://github.com/swvl/express-versioned-route/commit/c939a1ea56ceb96982139cd35c7ab2f3f4a5b8e3))


### BREAKING CHANGES

* **package manager:** add `@swvl:registry=https://npm.pkg.github.com` to your .npmrc

## [2.2.4](https://github.com/swvl/express-versioned-route/compare/v2.2.3...v2.2.4) (2021-12-02)


### Bug Fixes

* package.json & package-lock.json to reduce vulnerabilities ([c47146d](https://github.com/swvl/express-versioned-route/commit/c47146d4c96440902f04bb33f21e5e17baa4eaca))
* upgrade eslint-config-airbnb-base from 14.0.0 to 14.1.0 ([a212e99](https://github.com/swvl/express-versioned-route/commit/a212e990ccf0d5eb2e00269883bed6ca3fb2ee82))
* upgrade eslint-config-airbnb-base from 14.0.0 to 14.1.0 ([77acf46](https://github.com/swvl/express-versioned-route/commit/77acf46f16aee3a6368239cdae5b8c71d52b59cc))

## [2.2.3](https://github.com/swvl/express-versioned-route/compare/v2.2.2...v2.2.3) (2020-04-08)


### Bug Fixes

* **release:** change release rule to match cz format ([4c70df3](https://github.com/swvl/express-versioned-route/commit/4c70df386c0026fe42d9b1d4c29fba0b561b47bb))

## [2.2.2](https://github.com/swvl/express-versioned-route/compare/v2.2.1...v2.2.2) (2020-04-08)


### Bug Fixes

* **core:** function call should not pass extra arguments ([af4a447](https://github.com/swvl/express-versioned-route/commit/af4a4473d9c3d45075640843e397c03fbcefaab7))

## [2.2.1](https://github.com/swvl/express-versioned-route/compare/v2.2.0...v2.2.1) (2020-04-02)

# [2.2.0](https://github.com/swvl/express-versioned-route/compare/v2.1.1...v2.2.0) (2020-04-01)


### Features

* support custom headers for client selection ([adc382b](https://github.com/swvl/express-versioned-route/commit/adc382b724df3408fdd7c9cee9d3d7b96eb31931))

## [2.1.1](https://github.com/swvl/express-versioned-route/compare/v2.1.0...v2.1.1) (2020-03-23)


### Bug Fixes

* **release:** introduce a patch release on docs(README) & refactor commits ([775da4b](https://github.com/swvl/express-versioned-route/commit/775da4beb989116cde50e53cad80016f32cd4112))

# [2.1.0](https://github.com/swvl/express-versioned-route/compare/v2.0.0...v2.1.0) (2020-03-16)


### Bug Fixes

* soft deprecation ([6e2f405](https://github.com/swvl/express-versioned-route/commit/6e2f40592021205692972ad936057b423f7d2ffe)), closes [#4](https://github.com/swvl/express-versioned-route/issues/4)


### Features

* add support for a default version ([0f193ab](https://github.com/swvl/express-versioned-route/commit/0f193ab4add6b678f637c1792457a099edea5597)), closes [#9](https://github.com/swvl/express-versioned-route/issues/9)
* add support for lib configuration ([567723d](https://github.com/swvl/express-versioned-route/commit/567723deadd19ce58655222fdf470f695bbfa14f)), closes [#23](https://github.com/swvl/express-versioned-route/issues/23)
* add support for multiple clients ([d533fba](https://github.com/swvl/express-versioned-route/commit/d533fba7b001ac7b7e317f79f8a6278c37c0f268)), closes [#21](https://github.com/swvl/express-versioned-route/issues/21)

# [2.0.0](https://github.com/swvl/express-versioned-route/compare/v1.1.3...v2.0.0) (2020-03-11)


### Features

* refactor versioning headers ([d67ed6f](https://github.com/swvl/express-versioned-route/commit/d67ed6f4a1463cac7f85da5a7f58adbd68b1083a))


### BREAKING CHANGES

* client headers has been changed

## [1.1.3](https://github.com/swvl/express-versioned-route/compare/v1.1.2...v1.1.3) (2020-03-09)


### Bug Fixes

* run-kit location ([6dad532](https://github.com/swvl/express-versioned-route/commit/6dad532032477c89f432770222ec2d6b4ed25747))

## [1.1.2](https://github.com/swvl/express-versioned-route/compare/v1.1.1...v1.1.2) (2020-03-09)


### Bug Fixes

* versionSelectorMiddleware ([5a39f2a](https://github.com/swvl/express-versioned-route/commit/5a39f2aba94ba2b5cb6e94e5b73978c8c642cabb))

## [1.1.1](https://github.com/swvl/express-versioned-route/compare/v1.1.0...v1.1.1) (2020-03-08)


### Bug Fixes

* runkit example ([3a76631](https://github.com/swvl/express-versioned-route/commit/3a7663138c5603e15fa5c1c4c3d464e192ee17d3))

# [1.1.0](https://github.com/swvl/express-versioned-route/compare/v1.0.0...v1.1.0) (2020-03-08)


### Bug Fixes

* version selector middelware ([6440e09](https://github.com/swvl/express-versioned-route/commit/6440e0969b903138698462c374d0b8ca77322658))


### Features

* move example1 to run inside RunKit ([a00eda9](https://github.com/swvl/express-versioned-route/commit/a00eda979b2dac1f4a40479acf3aef1bf006ee87)), closes [#8](https://github.com/swvl/express-versioned-route/issues/8)

# [1.1.0-next.2](https://github.com/swvl/express-versioned-route/compare/v1.1.0-next.1...v1.1.0-next.2) (2020-03-08)


### Bug Fixes

* version selector middelware ([7426004](https://github.com/swvl/express-versioned-route/commit/7426004f3d6860f5dea501b98fe5dcb59fbb0a1d))

# [1.1.0-next.1](https://github.com/swvl/express-versioned-route/compare/v1.0.0...v1.1.0-next.1) (2020-03-08)


### Features

* move example1 to run inside RunKit ([63a9d12](https://github.com/swvl/express-versioned-route/commit/63a9d12e27071a9cc85fcd04c3faa1b1b4c85530)), closes [#8](https://github.com/swvl/express-versioned-route/issues/8)

# 1.0.0 (2020-03-05)


### Features


runkit-example.js
const app = require('@runkit/runkit/express-endpoint/1.0.0')(exports); // Replace this line by a normal express app
const { versionsDef } = require('@swvl/express-versioned-route');

const process = (name, req) => {
  if (!req.processingOrder) {
    req.processingOrder = [];
  }
  req.processingOrder.push(name);
};
const mwDef = name => (req, res, next) => {
  process(name, req);
  next();
};
const searchHandler = version => (req, res) => {
  process(version, req);
  res.status(200).send(JSON.stringify(req.processingOrder));
};
const mw1 = mwDef('mw1');
const mw2 = mwDef('mw2');
const searchMW1 = mwDef('searchMW1');
const searchMW2 = mwDef('searchMW2');
const dynamicSearchMW1 = mwDef('dynamicSearchMW1');
const dynamicSearchMW2 = mwDef('dynamicSearchMW2');
const deepSearchMW = mwDef('deepSearchMW');
const searchHandlerV1 = searchHandler('v1');
const searchHandlerV2 = searchHandler('v2');
const searchHandlerV3 = searchHandler('v3');
const searchHandlerV4 = searchHandler('v4');

const searchVersionDef = versionsDef()({
  versions: {
    simpleSearch: ['2020-Q1', searchMW1, searchMW2, searchHandlerV1],
    dynamicSearch: [dynamicSearchMW1, dynamicSearchMW2, searchHandlerV2, 'default'],
    superSearch: [searchHandlerV3],
    deepSearch: [deepSearchMW, searchHandlerV4],
  },
  Android: [
    [400, 'simpleSearch'],
    [450, 'dynamicSearch'],
    [500, 'superSearch'],
  ],
  iOS: [
    [400, 'simpleSearch'],
    [460, 'dynamicSearch'],
  ],
});

app.get('/', mw1, mw2, searchVersionDef);

process.env.RUNKIT_ENDPOINT_URL;
// curl -L -H 'accept-version: supersearch' <runkit-url>
